src/iterate_matrix.F

```diff
@@ -1072,6 +1078,228 @@ SUBROUTINE matrix_sign_proot(matrix_sign, matrix, threshold, sign_order)
+         ELSE IF (W(i) < 0) THEN
+            sm_sign(i, i) = -1.0
+         ELSE
+            sm_sign(i, i) = 0.0
+         ENDIF
+      ENDDO
+
+      ! Create matrix with eigenvalues in {-1,0,1} and eigenvectors of sm:
+      ! sm_sign = U * sm_sign * U.T
+      CALL dgemm('N', 'N', N, N, N, 1.0_dp, U, N, sm_sign, N, 0.0_dp, tmp, N)
+      CALL dgemm('N', 'T', N, N, N, 1.0_dp, tmp, N, U, N, 0.0_dp, sm_sign, N)
+
+      DEALLOCATE (W)
+      DEALLOCATE (tmp)
+      DEALLOCATE (U)
+   END SUBROUTINE dense_matrix_sign_direct
+
+! **************************************************************************************************
+!> \brief Submatrix method
+!> \param matrix_sign ...
+!> \param matrix ...
+!> \param threshold ...
+!> \param sign_order ...
+!> \param submatrix_sign_method ...
+!> \par History
+!>       2019.03 created [Robert Schade]
+!>       2019.06 impl. submatrix method [Michael Lass]
+!> \author Robert Schade, Michael Lass
+! **************************************************************************************************
+   SUBROUTINE matrix_sign_submatrix(matrix_sign, matrix, threshold, sign_order, submatrix_sign_method)
+
+      TYPE(dbcsr_type), INTENT(INOUT)                    :: matrix_sign, matrix
+      REAL(KIND=dp), INTENT(IN)                          :: threshold
+      INTEGER, INTENT(IN), OPTIONAL                      :: sign_order
+      INTEGER, INTENT(IN)                                :: submatrix_sign_method
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'matrix_sign_submatrix', &
+         routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: group, handle, i, myrank, nblkcols, &
+                                                            order, sm_size, unit_nr
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: my_sms
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :)        :: sm, sm_sign
+      TYPE(cp_logger_type), POINTER                      :: logger
+      TYPE(dbcsr_distribution_type)                      :: dist
+      TYPE(submatrix_dissection_type)                    :: dissection
+
+      CALL timeset(routineN, handle)
+
+      logger => cp_get_default_logger()
+      IF (logger%para_env%mepos == logger%para_env%source) THEN
+         unit_nr = cp_logger_get_default_unit_nr(logger, local=.TRUE.)
+      ELSE
+         unit_nr = -1
+      ENDIF
+
+      IF (PRESENT(sign_order)) THEN
+         order = sign_order
+      ELSE
+         order = 2
+      ENDIF
+
+      CALL dbcsr_get_info(matrix=matrix, nblkcols_total=nblkcols, distribution=dist, group=group)
+      CALL dbcsr_distribution_get(dist=dist, mynode=myrank)
+
+      CALL dissection%init(matrix)
+      CALL dissection%get_sm_ids_for_rank(myrank, my_sms)
+
+      !$OMP PARALLEL PRIVATE(sm, sm_sign, sm_size) SHARED(dissection, myrank, my_sms)
+      !$OMP DO SCHEDULE(GUIDED)
+      DO i = 1, SIZE(my_sms)
+         PRINT *, "Rank", myrank, "processing submatrix", my_sms(i)
```

Comment:
this bare print statement should probably be converted to using a logger facility